source/Applications/Basic/Visualization/ReadPCLVis3D/ReadPCLVis3D.cpp

```diff
@@ -1,35 +1,86 @@
+    {
+        viewer->addPointCloud<pcl::PointXYZRGBNormal>(pointCloud);
+
+        const int normalsSkipped = 10;
+        std::cout << "Note! 1 out of " << normalsSkipped << " normals are visualized" << std::endl;
+        viewer->addPointCloudNormals<pcl::PointXYZRGBNormal, pcl::PointXYZRGBNormal>(pointCloud,
+                                                                                     pointCloud,
+                                                                                     normalsSkipped,
+                                                                                     1,
+                                                                                     "normals");
+    }
+
+    void addPointCloudToViewer(boost::shared_ptr<pcl::visualization::PCLVisualizer> &viewer,
+                               const boost::shared_ptr<pcl::PointCloud<pcl::PointXYZRGB>> &pointCloud)
+    {
+        viewer->addPointCloud<pcl::PointXYZRGB>(pointCloud);
+    }
+
+    template<typename T>
+    void visualizePointCloudPCL(const boost::shared_ptr<pcl::PointCloud<T>> &pointCloud)
+    {
+        auto viewer = boost::make_shared<pcl::visualization::PCLVisualizer>("Viewer");
+
+        addPointCloudToViewer(viewer, pointCloud);
+
+        viewer->setCameraPosition(0, 0, -100, 0, -1, 0);
+
+        std::cout << "Press r to centre and zoom the viewer so that the entire cloud is visible" << std::endl;
+        std::cout << "Press q to exit the viewer application" << std::endl;
+        while(!viewer->wasStopped())
+        {
+            viewer->spinOnce(100);
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+    }
+
+} // namespace
+
+int main(int argc, char **argv)
 {
     try
     {
-        pcl::PointCloud<pcl::PointXYZRGB>::Ptr cloudPtr(new pcl::PointCloud<pcl::PointXYZRGB>);
+        bool useNormals = argc >= 2 && std::string(argv[1]) == "normals";
 
         std::string pointCloudFile = std::string(ZIVID_SAMPLE_DATA_DIR) + "/Zivid3D.pcd";
         std::cout << "Reading PCD point cloud from file: " << pointCloudFile << std::endl;
-        pcl::io::loadPCDFile<pcl::PointXYZRGB>(pointCloudFile, *cloudPtr);
-        std::cout << "Loaded " << cloudPtr->width * cloudPtr->height << " points" << std::endl;
 
-        std::cout << "Visualizing point cloud" << std::endl;
-        pcl::visualization::CloudViewer cloudViewer("Simple Cloud Viewer");
-        std::cout << "Running visualizer. Blocking until window closes" << std::endl;
-        cloudViewer.showCloud(cloudPtr);
-        std::cout << "Press r to centre and zoom the viewer so that the entire cloud is visible" << std::endl;
-        std::cout << "Press q to me exit the viewer application" << std::endl;
-        while(!cloudViewer.wasStopped())
+        if(useNormals)
         {
+            const auto pointCloudWithNormalsPCL = boost::make_shared<pcl::PointCloud<pcl::PointXYZRGBNormal>>();
```

Comment:
I don't think this is intended to be const in the way that you write the sample?
You load pcd file data into as the data of the pointer just below?